src/device_memory_tracker.cc

```diff
@@ -0,0 +1,281 @@
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "triton/backend/device_memory_tracker.h"
+
+#include <iostream>
+#include <stdexcept>
+#include "triton/core/tritonserver.h"
+
+namespace triton { namespace backend {
+
+std::unique_ptr<DeviceMemoryTracker> DeviceMemoryTracker::tracker_{nullptr};
+// Boilerplate from CUPTI examples
+namespace {
+
+// [WIP] figure out recover policy at different location of calling CUPTI
+#define CUPTI_CALL(call)                                      \
+  do {                                                        \
+    CUptiResult _status = call;                               \
+    if (_status != CUPTI_SUCCESS) {                           \
+      const char* errstr;                                     \
+      cuptiGetResultString(_status, &errstr);                 \
+      LOG_ERROR << #call << " failed with error: " << errstr; \
+      exit(EXIT_FAILURE);                                     \
+    }                                                         \
+  } while (0)
+
+#define BUF_SIZE (32 * 1024)
+#define ALIGN_SIZE (8)
+#define ALIGN_BUFFER(buffer, align)                                 \
+  (((uintptr_t)(buffer) & ((align)-1))                              \
+       ? ((buffer) + (align) - ((uintptr_t)(buffer) & ((align)-1))) \
+       : (buffer))
+
+void
+bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords)
+{
+  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
+  if (bfr != nullptr) {
+    *size = BUF_SIZE;
+    *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
+  }
+}
+
+void
+bufferCompleted(
+    CUcontext ctx, uint32_t streamId, uint8_t* buffer, size_t size,
+    size_t validSize)
+{
+  CUptiResult status;
+  CUpti_Activity* record = nullptr;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(buffer, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+        DeviceMemoryTracker::TrackActivity(record);
+      } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED)
+        break;
+      else {
+        CUPTI_CALL(status);
+      }
+    } while (1);
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
```

Comment:
protect with `if (buffer)` ?

Reply:
I believe free(0) is guranteed to be safe -

Reply:
Today I learned!